### PR TITLE
[mtouch/mmp] Share numerous command-line arguments.

### PR DESCRIPTION
**Pull request by rolfbjarne**
* Share --targetver, and deprecate mmp's --minos.
* Share --reference, and deprecate -r|--ref and -a|--assembly.
* Share --debug.
* Share --sdk.
* Share --warn-on-type-ref.
* Share --registrar and --runregistrar.
* Share --xml.
* Share --i18n.
* Share --nolink, --linksdkonly, --linkplatform and --linkskip.
* Share --cache.
* Share --force.
* Share --tls-provider and --http-message-handler.

This PR is probably easiest to reviewed commit-by-commit.

### CODE REVIEW

File: tools/common/Driver.cs [hunk 1]
```diff
@@ -67,6 +67,15 @@ static bool ParseOptions (Application app, Mono.Options.OptionSet options, strin
...
 			});
+			options.Add ("reference=", "Add an assembly to be processed.", v => app.References.Add (v));
+			// Unfortunately -r is used in mmp for something else (--resource), which means we can't use the same arguments for both mtouch and mmp.
+			// So add --reference, which is now used by both (and accepted by bgen as well), and deprecate -r|--ref for mtouch and -a|--assembly for mmp.
+#if MTOUCH
+			options.Add ("r|ref=", "Add an assembly to the resolver [DEPRECATED, use --reference instead]", v => app.References.Add (v), true);
+#endif
+#if MMP
+			options.Add ("a|assembly=", "Add an assembly to be processed [DEPRECATED, use --reference instead]", v => app.References.Add (v), true);
+#endif
```
Comment (spouliot) [inline tools/common/Driver.cs:78]: ^ should not those remain the the tool-specific files to avoid `#if`?

File: tools/common/Driver.cs [hunk 2]
```diff
@@ -75,6 +75,22 @@ static bool ParseOptions (Application app, Mono.Options.OptionSet options, strin
...
+				}
+			});
+#if MMP
+			options.Add ("minos=", "Minimum supported version of Mac OS X [DEPRECATED, use --targetver instead]", v => {
+				try {
+					app.DeploymentTarget = StringUtils.ParseVersion (v);
+				} catch (Exception ex) {
+					throw ErrorHelper.CreateError (26, ex, Errors.MX0026, $"minos:{v}", ex.Message);
+				}
+			}, true);
```
Comment (spouliot) [inline tools/common/Driver.cs:93]: same, why not keep it inside mmp itself ?

File: tools/common/Driver.cs (hunk 1 above)
Comment (dalexsoto) [reply tools/common/Driver.cs:78]: I guess it is fine to keep together and maybe be able to remove/hide in the future

Comment (rolfbjarne) [reply tools/common/Driver.cs:78]: @spouliot Yes, the idea is to keep the related code close, but I'm fine either way.

File: tools/common/Driver.cs [hunk 3]
```diff
@@ -47,17 +52,69 @@ static bool ParseOptions (Application app, Mono.Options.OptionSet options, strin
 			Action a = Action.None; // Need a temporary local variable, since anonymous functions can't write directly to ref/out arguments.
 
 			options.Add ("h|?|help", "Displays the help", v => a = Action.Help);
+			options.Add ("f|force", "Forces the recompilation of code, regardless of timestamps", v => Force = true);
+			options.Add ("cache=", "Specify the directory where temporary build files will be cached", v => app.Cache.Location = v);
 			options.Add ("version", "Output version information and exit.", v => a = Action.Version);
 			options.Add ("v|verbose", "Specify how verbose the output should be. This can be passed multiple times to increase the verbosity.", v => Verbosity++);
 			options.Add ("q|quiet", "Specify how quiet the output should be. This can be passed multiple times to increase the silence.", v => Verbosity--);
+			options.Add ("debug:", "Build a debug app. If AOT-compiling, will also generate native debug code for the specified assembly (set to 'all' to generate debug code for all assemblies, the default is to generate debug code for user assemblies only)", v => {
```
Comment (stephen-hawley) [inline tools/common/Driver.cs:60]: nit: the last sentence needs a `.` at the end.

File: tools/common/Driver.cs [hunk 4]
```diff
@@ -47,17 +52,69 @@ static bool ParseOptions (Application app, Mono.Options.OptionSet options, strin
...
+			options.Add ("minos=", "Minimum supported version of Mac OS X [DEPRECATED, use --targetver instead]", v => {
+				try {
+					app.DeploymentTarget = StringUtils.ParseVersion (v);
+				} catch (Exception ex) {
+					throw ErrorHelper.CreateError (26, ex, Errors.MX0026, $"minos:{v}", ex.Message);
+				}
+			}, true);
+#endif
 			options.Add ("sdkroot=", "Specify the location of Apple SDKs, default to 'xcode-select' value.", v => sdk_root = v);
+			options.Add ("sdk=", "Specifies the SDK version to compile against (version, for example \"10.9\")", v => {
```
Comment (stephen-hawley) [inline tools/common/Driver.cs:96]: needs a `.` at the end.

File: tools/common/Driver.cs [hunk 5]
```diff
@@ -47,17 +52,69 @@ static bool ParseOptions (Application app, Mono.Options.OptionSet options, strin
...
+				}
+			});
 			options.Add ("target-framework=", "Specify target framework to use. Currently supported: '" + string.Join ("', '", TargetFramework.ValidFrameworks.Select ((v) => v.ToString ())) + "'.", v => SetTargetFramework (v));
 #if MMP
 			options.Add ("abi=", "Comma-separated list of ABIs to target. x86_64", v => app.ParseAbi (v));
 #else
 			options.Add ("abi=", "Comma-separated list of ABIs to target. Currently supported: armv7, armv7+llvm, armv7+llvm+thumb2, armv7s, armv7s+llvm, armv7s+llvm+thumb2, arm64, arm64+llvm, arm64_32, arm64_32+llvm, i386, x86_64", v => app.ParseAbi (v));
 #endif
 			options.Add ("no-xcode-version-check", "Ignores the Xcode version check.", v => { min_xcode_version = null; }, true /* This is a non-documented option. Please discuss any customers running into the xcode version check on the maciosdev@ list before giving this option out to customers. */);
+			options.Add ("nolink", "Do not link the assemblies", v => app.LinkMode = LinkMode.None);
```
Comment (stephen-hawley) [inline tools/common/Driver.cs:110]: same

File: tools/common/Driver.cs [hunk 6]
```diff
@@ -202,6 +261,56 @@ static bool ParseOptions (Application app, Mono.Options.OptionSet options, strin
...
+				case "default":
+					app.Registrar = RegistrarMode.Default;
+					break;
+#if MMP
+				case "partial":
+				case "partial-static":
+					app.Registrar = RegistrarMode.PartialStatic;
+					break;
+				case "il":
+					app.Registrar = RegistrarMode.Dynamic;
```
Comment (stephen-hawley) [inline tools/common/Driver.cs:285]: Is `il` really an exact alias for dynamic? Does it need to be deprecated?

File: tools/common/Driver.cs [hunk 7]
```diff
@@ -202,6 +261,56 @@ static bool ParseOptions (Application app, Mono.Options.OptionSet options, strin
 					app.EnvironmentVariables.Add (name, value);
 				}
 			);
+			options.Add ("registrar:", "Specify the registrar to use (dynamic, static or default (dynamic in the simulator, static on device))", v => {
+				var split = v.Split ('=');
+				var name = split [0];
+				var value = split.Length > 1 ? split [1] : string.Empty;
```
Comment (mandel-macaque) [inline tools/common/Driver.cs:267]: Do we trust people not doing 'stupid' things like Static etc..?

File: tools/common/Driver.cs (hunk 6 above)
Comment (mandel-macaque) [reply tools/common/Driver.cs:285]: building on that comment, we can move the case next to the dynamic one.

File: tools/common/Driver.cs (hunk 7 above)
Comment (rolfbjarne) [reply tools/common/Driver.cs:251]: They'll just get an error saying to use `static` instead.

File: tools/common/Driver.cs (hunk 3 above)
Comment (rolfbjarne) [reply tools/common/Driver.cs]: @stephen-hawley fixed

File: tools/common/Driver.cs (hunk 4 above)
Comment (rolfbjarne) [reply tools/common/Driver.cs]: @stephen-hawley fixed

File: tools/common/Driver.cs (hunk 5 above)
Comment (rolfbjarne) [reply tools/common/Driver.cs]: @stephen-hawley fixed

File: tools/common/Driver.cs (hunk 6 above)
Comment (rolfbjarne) [reply tools/common/Driver.cs]: I just removed the `il` option.

File: tools/common/Driver.cs (hunk 1 above)
Comment (rolfbjarne) [reply tools/common/Driver.cs]: I've moved this code back into mtouch.cs/driver.cs

File: tools/common/Driver.cs (hunk 6 above)
Comment (stephen-hawley) [reply tools/common/Driver.cs]: 👍

**Pull request merged**